src/uct/rocm/rocm_common.c

```diff
@@ -0,0 +1,261 @@
+        uct_rocm_cfg.agents.gpu_info[uct_rocm_cfg.num_of_gpu].func = bdfid & 0x7;
+
+        ucs_debug("Found GPU agent : 0x%lx. [ B#%02d, D#%02d, F#%02d ]",
+                uct_rocm_cfg.agents.gpu_agent[uct_rocm_cfg.num_of_gpu].handle,
+                uct_rocm_cfg.agents.gpu_info[uct_rocm_cfg.num_of_gpu].bus,
+                uct_rocm_cfg.agents.gpu_info[uct_rocm_cfg.num_of_gpu].device,
+                uct_rocm_cfg.agents.gpu_info[uct_rocm_cfg.num_of_gpu].func);
+
+
+        uct_rocm_cfg.agents.gpu_info[uct_rocm_cfg.num_of_gpu].pool.handle
+                                                            = (uint64_t) -1;
+        status = hsa_amd_agent_iterate_memory_pools(agent,
+                                    uct_rocm_hsa_amd_memory_pool_callback,
+                                    &uct_rocm_cfg.agents.gpu_info[uct_rocm_cfg.num_of_gpu].pool);
+
+        if (status != HSA_STATUS_SUCCESS && status != HSA_STATUS_INFO_BREAK) {
+            ucs_error("Failure to iterate regions: 0x%x\n", status);
+            return status;
+        }
+
+        if (uct_rocm_cfg.agents.gpu_info[uct_rocm_cfg.num_of_gpu].pool.handle
+                                                    == (uint64_t)-1) {
+            ucs_warn("Could not find memory pool for GPU agent");
+        }
+
+        uct_rocm_cfg.num_of_gpu++;
+
+    } else  if (device_type == HSA_DEVICE_TYPE_CPU) {
+        uct_rocm_cfg.agents.cpu.agent = agent;
+        ucs_debug("Found CPU agent : 0x%lx", uct_rocm_cfg.agents.cpu.agent.handle);
+
+        uct_rocm_cfg.agents.cpu.pool.handle = (uint64_t) -1;
+        status = hsa_amd_agent_iterate_memory_pools(agent,
+                                                    uct_rocm_hsa_amd_memory_pool_callback,
+                                                    &uct_rocm_cfg.agents.cpu.pool);
+
+        if (status != HSA_STATUS_SUCCESS && status != HSA_STATUS_INFO_BREAK) {
+            ucs_error("Failure to iterate memory pools: 0x%x", status);
+            return status;
+        }
+
+        if (uct_rocm_cfg.agents.cpu.pool.handle == (uint64_t)-1) {
+            ucs_warn("Could not find memory pool for CPU agent");
+        }
+    }
+
+    /* Keep iterating */
+    return HSA_STATUS_SUCCESS;
+}
+
+
+hsa_status_t uct_rocm_init()
+{
+    hsa_status_t status;
+
+    if (pthread_mutex_lock(&rocm_init_mutex) == 0) {
+        if (rocm_ucx_initialized) {
+            status =  HSA_STATUS_SUCCESS;
+            goto end;
+        }
+    } else  {
+        ucs_error("Could not take mutex");
+        status = HSA_STATUS_ERROR;
+        return status;
+    }
+
+    /* Initialize HSA RT just in case if it was not initialized before */
+    status = hsa_init();
+
+    if (status != HSA_STATUS_SUCCESS) {
+        ucs_error("Failure to open HSA connection: 0x%x", status);
+        goto end;
+    }
+
+    /* Collect information about GPU agents */
+    status = hsa_iterate_agents(uct_rocm_hsa_agent_callback, NULL);
```

Comment:
this is internal rocm design decision, but i would recommend that for simplicity there is no need to cache the list of devices.